src/rust/apps/ethereum/src/address.rs

```diff
@@ -0,0 +1,119 @@
+use sha3::digest::Digest;
+
+/// Generates a checksummed ethereum hex address from a 20 byte recipient.
+/// `recipient` - 20 byte tail (last 20 bytes of the pubkeyhash).
+/// `result` out will hold the address string, must have space for 42 chars.
+pub fn from_pubkey_hash<T>(recipient: &[u8; 20], result: &mut T) -> core::fmt::Result
+where
+    T: core::fmt::Write,
+{
+    let mut hex = [0u8; 40];
+    hex::encode_to_slice(recipient, &mut hex).unwrap();
+    let hash = sha3::Keccak256::digest(&hex[..]);
+    for (i, e) in hex.iter_mut().enumerate() {
+        let hash_byte = {
+            let b = hash[i / 2];
+            if i % 2 == 0 {
+                b >> 4
+            } else {
+                b & 0xf
+            }
+        };
+        if *e > b'9' && hash_byte > 7 {
+            *e -= 32; // convert to uppercase
+        }
+    }
+    write!(result, "0x{}", unsafe {
+        // valid utf8 because hex and the uppercasing above is correct.
+        core::str::from_utf8_unchecked(&hex[..])
+    })
+}
+
+/// Generates a checksummed ethereum hex address from a 65 byte pubkey.
+/// `recipient` - 20 byte tail (last 20 bytes of the pubkeyhash).
+/// `result` out will hold the address string, must have space for 42 chars.
+pub fn from_pubkey<T>(pubkey_uncompressed: &[u8; 65], result: &mut T) -> core::fmt::Result
+where
+    T: core::fmt::Write,
+{
+    let hash = sha3::Keccak256::digest(&pubkey_uncompressed[1..]);
+    let mut last20: [u8; 20] = [0; 20];
+    last20.copy_from_slice(&hash[hash.len() - 20..]);
+    from_pubkey_hash(&last20, result)
```

Comment:
i don't like that i can't get to a array from a slice directly without copying :/

Reply:
What you want is `const generics` which we don't have yet. But we have workarounds ;)!, if you add the "arrayref" crate you can write the following:

```rust
    // Would have preferred sha3::Keccak256::output_size(), but it isn't const fn.
    let last20 = arrayref::array_ref!(hash, 32-20, 20);
```

Eventually (probably in a year or two) we will have proper const generics and all these crates will go away.

Reply:
Hah, actually `hash.len()` is const so the following works:

```rust
    let last20 = arrayref::array_ref!(hash, hash.len()-20, 20);
```

Reply:
excellent! this is much better.